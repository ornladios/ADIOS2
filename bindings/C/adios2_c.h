/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * adios2_c.h : ADIOS2 C bindings declarations
 *
 *  Created on: Mar 13, 2017
 *      Author: William F Godoy godoywf@ornl.gov
 */

#ifndef ADIOS2_BINDINGS_C_ADIOS2_C_H_
#define ADIOS2_BINDINGS_C_ADIOS2_C_H_

#include "adios2/ADIOSConfig.h"

#include "c/adios2_c_adios.h"
#include "c/adios2_c_attribute.h"
#include "c/adios2_c_engine.h"
#include "c/adios2_c_io.h"
#include "c/adios2_c_operator.h"
#include "c/adios2_c_variable.h"

#endif /* ADIOS2_BINDINGS_C_ADIOS2_C_H_ */
