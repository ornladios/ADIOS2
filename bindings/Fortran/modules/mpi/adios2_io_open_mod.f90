!
! Distributed under the OSI-approved Apache License, Version 2.0.  See
!  accompanying file Copyright.txt for details.
!
!  adios2_io_open_mod.f90 : ADIOS2 Fortran bindings for IO class open function
!
!   Created on: Mar 13, 2017
!       Author: William F Godoy godoywf@ornl.gov
!

module adios2_io_open_mod
    use adios2_parameters_mod
    implicit none

    interface adios2_open
        module procedure adios2_open_old_comm
        module procedure adios2_open_new_comm
    end interface

contains

    subroutine adios2_open_old_comm(engine, io, name, adios2_mode, ierr)
        type(adios2_engine), intent(out) :: engine
        type(adios2_io), intent(in) :: io
        character*(*), intent(in) :: name
        integer, intent(in) :: adios2_mode
        integer, intent(out) :: ierr

        engine%mode = adios2_mode
        call adios2_open_f2c(engine%f2c, io%f2c, TRIM(ADJUSTL(name))//char(0), &
                             adios2_mode, ierr)

        if( ierr == 0 ) then
            engine%valid = .true.
            engine%name = name
            call adios2_engine_get_type_f2c(engine%type, engine%f2c, ierr)
            engine%mode = adios2_mode
        end if

    end subroutine

    subroutine adios2_open_new_comm(engine, io, name, adios2_mode, comm, ierr)
        type(adios2_engine), intent(out) :: engine
        type(adios2_io), intent(in) :: io
        character*(*), intent(in) :: name
        integer, intent(in) :: adios2_mode
        integer, intent(in) :: comm
        integer, intent(out) :: ierr

        call adios2_open_new_comm_f2c(engine%f2c, io%f2c, &
                                      TRIM(ADJUSTL(name))//char(0), &
                                      adios2_mode, comm, ierr)

        if( ierr == 0 ) then
            engine%valid = .true.
            engine%name = name
            call adios2_engine_get_type_f2c(engine%type, engine%f2c, ierr)
            engine%mode = adios2_mode
        end if

    end subroutine

end module
