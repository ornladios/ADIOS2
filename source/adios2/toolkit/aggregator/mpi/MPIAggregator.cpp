/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * MPIAggregator.cpp
 *
 *  Created on: Feb 20, 2018
 *      Author: William F Godoy godoywf@ornl.gov
 */

#include "MPIAggregator.h"

#include "adios2/helper/adiosFunctions.h"

namespace adios2
{
namespace aggregator
{

MPIAggregator::MPIAggregator() : m_Comm(MPI_COMM_SELF) {}

MPIAggregator::~MPIAggregator()
{
    if (m_IsActive)
    {
        helper::CheckMPIReturn(MPI_Comm_free(&m_Comm),
                               "freeing aggregators comm in MPIAggregator "
                               "destructor, not recommended");
    }
}

void MPIAggregator::Init(const size_t subStreams, MPI_Comm parentComm) {}

void MPIAggregator::SwapBuffers(const int step) noexcept {}

void MPIAggregator::ResetBuffers() noexcept {}

BufferSTL &MPIAggregator::GetConsumerBuffer(BufferSTL &bufferSTL)
{
    return bufferSTL;
}

void MPIAggregator::Close()
{
    if (m_IsActive)
    {
        helper::CheckMPIReturn(MPI_Comm_free(&m_Comm),
                               "freeing aggregators comm at Close\n");
        m_IsActive = false;
    }
}

// PROTECTED
void MPIAggregator::InitComm(const size_t subStreams, MPI_Comm parentComm)
{
    int parentRank;
    int parentSize;
    MPI_Comm_rank(parentComm, &parentRank);
    MPI_Comm_size(parentComm, &parentSize);

    const size_t processes = static_cast<size_t>(parentSize);
    size_t stride = processes / subStreams + 1;
    const size_t remainder = processes % subStreams;

    size_t consumer = 0;

    for (auto s = 0; s < subStreams; ++s)
    {
        if (s >= remainder)
        {
            stride = processes / subStreams;
        }

        if (static_cast<size_t>(parentRank) >= consumer &&
            static_cast<size_t>(parentRank) < consumer + stride)
        {
            helper::CheckMPIReturn(
                MPI_Comm_split(parentComm, static_cast<int>(consumer),
                               parentRank, &m_Comm),
                "creating aggregators comm with split at Open");
            m_ConsumerRank = static_cast<int>(consumer);
            m_SubStreamIndex = static_cast<size_t>(s);
        }

        consumer += stride;
    }

    MPI_Comm_rank(m_Comm, &m_Rank);
    MPI_Comm_size(m_Comm, &m_Size);

    if (m_Rank != 0)
    {
        m_IsConsumer = false;
    }

    m_IsActive = true;
    m_SubStreams = subStreams;
}

void MPIAggregator::HandshakeRank(const int rank)
{
    int message = -1;
    if (m_Rank == rank)
    {
        message = m_Rank;
    }

    helper::CheckMPIReturn(MPI_Bcast(&message, 1, MPI_INT, rank, m_Comm),
                           "handshake with aggregator rank 0 at Open");
}

} // end namespace aggregator
} // end namespace adios2
