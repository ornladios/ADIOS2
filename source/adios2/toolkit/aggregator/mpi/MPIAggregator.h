/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * MPIAggregator.h
 *
 *  Created on: Feb 20, 2018
 *      Author: William F Godoy godoywf@ornl.gov
 */

#ifndef ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPIAGGREGATOR_H_
#define ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPIAGGREGATOR_H_

#include "adios2/ADIOSMPI.h"
#include "adios2/ADIOSTypes.h"
#include "adios2/toolkit/format/BufferSTL.h"

namespace adios2
{
namespace aggregator
{

class MPIAggregator
{
public:
    /** total number of substreams */
    size_t m_SubStreams = 0;

    /** current substream index from 0 to m_SubStreams-1 */
    size_t m_SubStreamIndex = 0;

    /** split Communicator for a substream: producers and consumer (rank=0) */
    MPI_Comm m_Comm;

    /** rank from m_Comm */
    int m_Rank = 0;

    /** size from m_Comm */
    int m_Size = 1;

    /**
     * true: consumes data from itself or other processes and interacts with
     * transport managers
     */
    bool m_IsConsumer = true;

    /** true: doing aggregation, false: not doing aggregation */
    bool m_IsActive = false;

    /** consumer original rank coming from the parent communicator in Init,
     *  corresponds to m_Rank = 0 */
    int m_ConsumerRank = -1;

    MPIAggregator();

    virtual ~MPIAggregator();

    virtual void Init(const size_t subStreams, MPI_Comm parentComm);

    virtual std::vector<std::vector<MPI_Request>>
    IExchange(BufferSTL &bufferSTL, const int step) = 0;

    virtual std::vector<std::vector<MPI_Request>>
    IExchangeAbsolutePosition(BufferSTL &bufferSTL, const int step) = 0;

    virtual void
    WaitAbsolutePosition(std::vector<std::vector<MPI_Request>> &requests,
                         const int step) = 0;

    virtual void Wait(std::vector<std::vector<MPI_Request>> &requests,
                      const int step) = 0;

    virtual void SwapBuffers(const int step) noexcept;

    virtual void ResetBuffers() noexcept;

    virtual BufferSTL &GetConsumerBuffer(BufferSTL &bufferSTL);

    /** closes current aggregator, frees m_Comm */
    void Close();

protected:
    /** Init m_Comm splitting assigning ranks to subStreams (balanced except for
     * the last rank) */
    void InitComm(const size_t subStreams, MPI_Comm parentComm);

    /** handshakes a single rank with the rest of the m_Comm ranks */
    void HandshakeRank(const int rank = 0);

    /** assigning extra buffers for aggregation */
    std::vector<BufferSTL> m_Buffers;
};

} // end namespace aggregator
} // end namespace adios2

#endif /* ADIOS2_TOOLKIT_AGGREGATOR_MPIAGGREGATOR_H_ */
