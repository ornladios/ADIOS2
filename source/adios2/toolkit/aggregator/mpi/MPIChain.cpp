/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * MPIChain.cpp
 *
 *  Created on: Feb 21, 2018
 *      Author: William F Godoy godoywf@ornl.gov
 */
#include "MPIChain.h"

#include "adios2/ADIOSMPI.h"
#include "adios2/helper/adiosFunctions.h" //helper::CheckMPIReturn

namespace adios2
{
namespace aggregator
{

MPIChain::MPIChain() : MPIAggregator() {}

void MPIChain::Init(const size_t subStreams, MPI_Comm parentComm)
{
    InitComm(subStreams, parentComm);
    HandshakeRank(0);
    HandshakeLinks();

    // add a receiving buffer except for the last rank (only sends)
    if (m_Rank < m_Size)
    {
        m_Buffers.emplace_back(); // just one for now
    }
}

std::vector<std::vector<MPI_Request>> MPIChain::IExchange(BufferSTL &bufferSTL,
                                                          const int step)
{
    if (m_Size == 1)
    {
        return std::vector<std::vector<MPI_Request>>();
    }

    BufferSTL &sendBuffer = GetSender(bufferSTL);
    const int endRank = m_Size - 1 - step;
    const bool sender = (m_Rank >= 1 && m_Rank <= endRank) ? true : false;
    const bool receiver = (m_Rank < endRank) ? true : false;

    std::vector<std::vector<MPI_Request>> requests(2);

    if (sender) // sender
    {
        requests[0].resize(1);

        helper::CheckMPIReturn(MPI_Isend(&sendBuffer.m_Position, 1,
                                         ADIOS2_MPI_SIZE_T, m_Rank - 1, 0,
                                         m_Comm, &requests[0][0]),
                               ", aggregation Isend size at iteration " +
                                   std::to_string(step) + "\n");

        // only send data if buffer larger than 0
        if (sendBuffer.m_Position > 0)
        {

            const std::vector<MPI_Request> requestsISend64 =
                helper::Isend64(sendBuffer.m_Buffer.data(),
                                sendBuffer.m_Position, m_Rank - 1, 1, m_Comm,
                                ", aggregation Isend64 data at iteration " +
                                    std::to_string(step));

            requests[0].insert(requests[0].end(), requestsISend64.begin(),
                               requestsISend64.end());
        }
    }
    // receive size, resize receiving buffer and receive data
    if (receiver)
    {
        size_t bufferSize = 0;
        MPI_Request receiveSizeRequest;
        helper::CheckMPIReturn(MPI_Irecv(&bufferSize, 1, ADIOS2_MPI_SIZE_T,
                                         m_Rank + 1, 0, m_Comm,
                                         &receiveSizeRequest),
                               ", aggregation Irecv size at iteration " +
                                   std::to_string(step) + "\n");

        MPI_Status receiveStatus;
        helper::CheckMPIReturn(
            MPI_Wait(&receiveSizeRequest, &receiveStatus),
            ", aggregation waiting for receiver size at iteration " +
                std::to_string(step) + "\n");

        BufferSTL &receiveBuffer = GetReceiver(bufferSTL);
        ResizeUpdateBufferSTL(
            bufferSize, receiveBuffer,
            "in aggregation, when resizing receiving buffer to size " +
                std::to_string(bufferSize));

        // only receive data if buffer is larger than 0
        if (bufferSize > 0)
        {
            requests[1] =
                helper::Irecv64(receiveBuffer.m_Buffer.data(),
                                receiveBuffer.m_Position, m_Rank + 1, 1, m_Comm,
                                ", aggregation Irecv64 data at iteration " +
                                    std::to_string(step));
        }
    }

    return requests;
}

std::vector<std::vector<MPI_Request>>
MPIChain::IExchangeAbsolutePosition(BufferSTL &bufferSTL, const int step)
{
    if (m_Size == 1)
    {
        return std::vector<std::vector<MPI_Request>>();
    }

    if (m_IsInExchangeAbsolutePosition)
    {
        throw std::runtime_error("ERROR: MPIChain::IExchangeAbsolutePosition: "
                                 "An existing exchange is still active.");
    }

    const int destination = (step != m_Size - 1) ? step + 1 : 0;
    std::vector<std::vector<MPI_Request>> requests(2,
                                                   std::vector<MPI_Request>(1));

    if (step == 0)
    {
        m_SizeSend =
            (m_Rank == 0) ? bufferSTL.m_AbsolutePosition : bufferSTL.m_Position;
    }

    if (m_Rank == step)
    {
        m_ExchangeAbsolutePosition =
            (m_Rank == 0) ? m_SizeSend
                          : m_SizeSend + bufferSTL.m_AbsolutePosition;

        helper::CheckMPIReturn(
            MPI_Isend(&m_ExchangeAbsolutePosition, 1, ADIOS2_MPI_SIZE_T,
                      destination, 0, m_Comm, &requests[0][0]),
            ", aggregation Isend absolute position at iteration " +
                std::to_string(step) + "\n");
    }
    else if (m_Rank == destination)
    {
        helper::CheckMPIReturn(
            MPI_Irecv(&bufferSTL.m_AbsolutePosition, 1, ADIOS2_MPI_SIZE_T, step,
                      0, m_Comm, &requests[1][0]),
            ", aggregation Irecv absolute position at iteration " +
                std::to_string(step) + "\n");
    }

    m_IsInExchangeAbsolutePosition = true;
    return requests;
}

void MPIChain::Wait(std::vector<std::vector<MPI_Request>> &requests,
                    const int step)
{
    if (m_Size == 1)
    {
        return;
    }

    const int endRank = m_Size - 1 - step;
    const bool sender = (m_Rank >= 1 && m_Rank <= endRank) ? true : false;
    const bool receiver = (m_Rank < endRank) ? true : false;

    MPI_Status status;
    if (receiver)
    {
        for (auto &req : requests[1])
        {
            helper::CheckMPIReturn(
                MPI_Wait(&req, &status),
                ", aggregation waiting for receiver request at iteration " +
                    std::to_string(step) + "\n");
        }
    }

    if (sender)
    {
        for (auto &req : requests[0])
        {
            helper::CheckMPIReturn(
                MPI_Wait(&req, &status),
                ", aggregation waiting for sender request at iteration " +
                    std::to_string(step) + "\n");
        }
    }
}

void MPIChain::WaitAbsolutePosition(
    std::vector<std::vector<MPI_Request>> &requests, const int step)
{
    if (m_Size == 1)
    {
        return;
    }

    if (!m_IsInExchangeAbsolutePosition)
    {
        throw std::runtime_error("ERROR: MPIChain::WaitAbsolutePosition: An "
                                 "existing exchange is not active.");
    }

    MPI_Status status;
    const int destination = (step != m_Size - 1) ? step + 1 : 0;

    if (m_Rank == destination)
    {
        helper::CheckMPIReturn(
            MPI_Wait(&requests[1][0], &status),
            ", aggregation Irecv Wait absolute position at iteration " +
                std::to_string(step) + "\n");
    }

    if (m_Rank == step)
    {
        helper::CheckMPIReturn(
            MPI_Wait(&requests[0][0], &status),
            ", aggregation Isend Wait absolute position at iteration " +
                std::to_string(step) + "\n");
    }
    m_IsInExchangeAbsolutePosition = false;
}

void MPIChain::SwapBuffers(const int /*step*/) noexcept
{
    m_CurrentBufferOrder = (m_CurrentBufferOrder == 0) ? 1 : 0;
}

void MPIChain::ResetBuffers() noexcept { m_CurrentBufferOrder = 0; }

BufferSTL &MPIChain::GetConsumerBuffer(BufferSTL &bufferSTL)
{
    return GetSender(bufferSTL);
}

// PRIVATE
void MPIChain::HandshakeLinks()
{
    int link = -1;

    MPI_Request sendRequest;
    if (m_Rank > 0) // send
    {
        helper::CheckMPIReturn(
            MPI_Isend(&m_Rank, 1, MPI_INT, m_Rank - 1, 0, m_Comm, &sendRequest),
            "Isend handshake with neighbor, MPIChain aggregator, at Open");
    }

    if (m_Rank < m_Size - 1) // receive
    {
        MPI_Request receiveRequest;
        helper::CheckMPIReturn(
            MPI_Irecv(&link, 1, MPI_INT, m_Rank + 1, 0, m_Comm,
                      &receiveRequest),
            "Irecv handshake with neighbor, MPIChain aggregator, at Open");

        MPI_Status receiveStatus;
        helper::CheckMPIReturn(
            MPI_Wait(&receiveRequest, &receiveStatus),
            "Irecv Wait handshake with neighbor, MPIChain aggregator, at Open");
    }

    if (m_Rank > 0)
    {
        MPI_Status sendStatus;
        helper::CheckMPIReturn(
            MPI_Wait(&sendRequest, &sendStatus),
            "Isend wait handshake with neighbor, MPIChain aggregator, at Open");
    }
}

BufferSTL &MPIChain::GetSender(BufferSTL &bufferSTL)
{
    if (m_CurrentBufferOrder == 0)
    {
        return bufferSTL;
    }
    else
    {
        return m_Buffers.front();
    }
}

BufferSTL &MPIChain::GetReceiver(BufferSTL &bufferSTL)
{
    if (m_CurrentBufferOrder == 0)
    {
        return m_Buffers.front();
    }
    else
    {
        return bufferSTL;
    }
}

void MPIChain::ResizeUpdateBufferSTL(const size_t newSize, BufferSTL &bufferSTL,
                                     const std::string hint)
{
    bufferSTL.Resize(newSize, hint);
    bufferSTL.m_Position = bufferSTL.m_Buffer.size();
}

} // end namespace aggregator
} // end namespace adios2
