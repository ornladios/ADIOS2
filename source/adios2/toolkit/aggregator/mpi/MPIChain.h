/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * MPIChain.h
 *
 *  Created on: Feb 20, 2018
 *      Author: William F Godoy godoywf@ornl.gov
 */

#ifndef ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPICHAIN_H_
#define ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPICHAIN_H_

#include "adios2/toolkit/aggregator/mpi/MPIAggregator.h"

namespace adios2
{
namespace aggregator
{

class MPIChain : public MPIAggregator
{

public:
    MPIChain();

    ~MPIChain() = default;

    void Init(const size_t subStreams, MPI_Comm parentComm) final;

    std::vector<std::vector<MPI_Request>> IExchange(BufferSTL &bufferSTL,
                                                    const int step) final;

    std::vector<std::vector<MPI_Request>>
    IExchangeAbsolutePosition(BufferSTL &bufferSTL, const int step) final;

    void Wait(std::vector<std::vector<MPI_Request>> &request,
              const int step) final;

    void WaitAbsolutePosition(std::vector<std::vector<MPI_Request>> &requests,
                              const int step) final;

    void SwapBuffers(const int step) noexcept final;

    void ResetBuffers() noexcept final;

    BufferSTL &GetConsumerBuffer(BufferSTL &bufferSTL) final;

private:
    bool m_IsInExchangeAbsolutePosition = false;
    size_t m_SizeSend = 0;
    size_t m_ExchangeAbsolutePosition = 0;

    /** current sender/receiver pair order (0: sender is original buffer and
    receiver is extra buffer, 1: sender is extra buffer and receiver is the
    original buffer)*/
    unsigned int m_CurrentBufferOrder = 0;

    void HandshakeLinks();

    /**
     * Returns a reference to the sender buffer depending on
     * m_CurrentBufferOrder
     * flag
     * @param bufferSTL original buffer from serializer
     * @return reference to sender buffer
     */
    BufferSTL &GetSender(BufferSTL &bufferSTL);

    /**
     * Returns a reference to the receiver buffer depending on
     * m_CurrentBufferOrder
     * flag
     * @param bufferSTL original buffer from serializer
     * @return reference to receiver buffer
     */
    BufferSTL &GetReceiver(BufferSTL &bufferSTL);

    /**
     * Resizes and updates m_Position in a bufferSTL, used for receiving buffers
     * @param newSize new size for receiving buffer
     * @param bufferSTL to be resized
     * @param hint used in exception error message
     */
    void ResizeUpdateBufferSTL(const size_t newSize, BufferSTL &bufferSTL,
                               const std::string hint);
};

} // end namespace aggregator
} // end namespace adios2

#endif /* ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPICHAIN_H_ */
