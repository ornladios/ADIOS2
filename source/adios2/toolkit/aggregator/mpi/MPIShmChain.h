/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * MPIShmChain.h
 *
 *  Created on: July 5, 2021
 *      Author: Norbert Podhorszki pnorbert@ornl.gov
 *
 */

#ifndef ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPISHMCHAIN_H_
#define ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPISHMCHAIN_H_

#include "adios2/common/ADIOSConfig.h"
#include "adios2/toolkit/aggregator/mpi/MPIAggregator.h"
#include "adios2/toolkit/shm/Spinlock.h"

#include <atomic>
#include <chrono>
#include <thread>

namespace adios2
{
namespace aggregator
{

// constexpr size_t SHM_BUF_SIZE = 4194304; // 4MB
// we allocate 2x this size + a bit for shared memory segment

/** A one- or two-layer aggregator chain for using Shared memory within a
 * compute node.
 * Use MPI split type to group processes within a node into one chain.
 * Depending on the number of aggregators, multiple nodes may be merged into
 * a two-layer chain, (multiple Aggregators and one Master).
 * Otherwise its a simple chain with one Aggregator=Master.
 *
 * m_Comm is the communicator that split for each node
 *
 */

class MPIShmChain : public MPIAggregator
{

public:
    MPIShmChain();

    ~MPIShmChain();

    /* Create a per-node communicator and return number of nodes */
    size_t PreInit(helper::Comm const &parentComm);

    void Init(const size_t numAggregators, const size_t subStreams,
              helper::Comm const &parentComm) final;

    void Close() final;

    /**
     * true: the Master (aggregator) process in the chain
     * always m_Rank == m_Comm.Rank() == 0 for a master aggregator
     * same as (m_AggregatorChainComm.Rank() == 0)
     */
    bool m_IsMasterAggregator = true;

    /*
        Variables set in PreInit
    */

    bool PreInitCalled = false;
    /* Communicator per compute node */
    helper::Comm m_NodeComm;
    /* Communicator connecting rank 0 on each node
       Useful only on rank 0s of m_NodeComm */
    helper::Comm m_OnePerNodeComm;
    /* Number of Compute Nodes
     * (size of m_OnePerNodeComm created from rank 0s of m_NodeComm)
     */
    size_t m_NumNodes;

    /*
        Variables set in Init
    */

    /* Communicator connecting all aggregators
        (rank 0 of each aggregator group)
        Useful only on aggregators themselves
    */
    helper::Comm m_AllAggregatorsComm;

    /* Communicator connecting the aggregators
       that write to the same substream.
       rank 0 becomes a MasterAggregator
       Useful only on aggregators themselves
       */
    helper::Comm m_AggregatorChainComm;

    struct ShmDataBuffer
    {
        size_t max_size;    // max size for buf
        size_t actual_size; // size of actual content
        // points to data buffer in shared memory
        // Warning: This is a different address on every process
        char *buf;
    };

    ShmDataBuffer *LockProducerBuffer();
    void UnlockProducerBuffer();
    ShmDataBuffer *LockConsumerBuffer();
    void UnlockConsumerBuffer();
    void ResetBuffers() noexcept;

    // 2*blocksize+some is allocated but only up to maxsegmentsize
    void CreateShm(size_t blocksize, const size_t maxsegmentsize,
                   const size_t alignment_size);
    void DestroyShm();

private:
    struct HandshakeStruct
    {
        int sendToken;
        int recvToken;
        helper::Comm::Req sendRequest;
        helper::Comm::Req recvRequest;
    };

    void HandshakeLinks_Start(helper::Comm &comm, HandshakeStruct &hs);
    void HandshakeLinks_Complete(HandshakeStruct &hs);

    helper::Comm::Win m_Win;

    enum class LastBufferUsed
    {
        None,
        A,
        B
    };

    struct ShmSegment
    {
        LastBufferUsed producerBuffer;
        LastBufferUsed consumerBuffer;
        unsigned int NumBuffersFull;
        // user facing structs
        ShmDataBuffer sdbA;
        ShmDataBuffer sdbB;
        shm::Spinlock lockSegment;
        // locks for individual buffers (sdb and buf)
        shm::Spinlock lockA;
        shm::Spinlock lockB;
        // the actual data buffers
        // char bufA[SHM_BUF_SIZE];
        // char bufB[SHM_BUF_SIZE];
    };
    ShmSegment *m_Shm;
    char *m_ShmBufA;
    char *m_ShmBufB;
};

} // end namespace aggregator
} // end namespace adios2

#endif /* ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPISHMCHAIN_H_ */
