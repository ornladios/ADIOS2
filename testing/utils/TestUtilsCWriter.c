/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * TestUtilsCWriter
 *
 *  Created on: Sept 19, 2018
 *      Author: Norbert Podhorszki
 */

#include <stdio.h>
#include <string.h>

#include <adios2_c.h>

#ifdef ADIOS2_HAVE_MPI
#include <mpi.h>
#endif

#include "SmallTestData_c.h"

int main(int argc, char *argv[])
{
    int rank = 0;
    int nproc = 1;
#ifdef ADIOS2_HAVE_MPI
    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &nproc);
    adios2_adios *adiosH = adios2_init(MPI_COMM_WORLD, adios2_debug_mode_on);
#else
    adios2_adios *adiosH = adios2_init(adios2_debug_mode_on);
#endif

    char engineName[32] = "BPFile";
    if (argc > 1)
    {
        strncpy(engineName, argv[1], sizeof(engineName));
    }

    // IO
    adios2_io *ioH = adios2_declare_io(adiosH, "CArrayTypes");
    // Set engine parameters
    adios2_set_engine(ioH, engineName);
    adios2_set_parameter(ioH, "ProfileUnits", "Microseconds");
    adios2_set_parameter(ioH, "Threads", "1");

    // 1D shape
    size_t shape[1], start[1], count[1];
    shape[0] = nproc * data_Nx;
    start[0] = rank * data_Nx;
    count[0] = data_Nx;

    // 2D shape
    size_t shape2[2], start2[2], count2[2];
    shape2[0] = nproc * d2_Nx;
    shape2[1] = d2_Ny;
    start2[0] = rank * d2_Nx;
    start2[1] = 0;
    count2[0] = d2_Nx;
    count2[1] = d2_Ny;

    // Define variables in ioH
    adios2_define_variable(ioH, "nproc", adios2_type_int32_t, 0, NULL, NULL,
                           NULL, adios2_constant_dims_true);

    adios2_define_variable(ioH, "varI8", adios2_type_int8_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varI16", adios2_type_int16_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varI32", adios2_type_int32_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varI64", adios2_type_int64_t, 1, shape, start,
                           count, adios2_constant_dims_true);

    adios2_define_variable(ioH, "varU8", adios2_type_uint8_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varU16", adios2_type_uint16_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varU32", adios2_type_uint32_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varU64", adios2_type_uint64_t, 1, shape, start,
                           count, adios2_constant_dims_true);

    adios2_define_variable(ioH, "varR32", adios2_type_float, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varR64", adios2_type_double, 1, shape, start,
                           count, adios2_constant_dims_true);

    adios2_define_variable(ioH, "R64_2d", adios2_type_double, 2, shape2, start2,
                           count2, adios2_constant_dims_true);

    // Define attributes in ioH
    adios2_define_attribute(ioH, "name", adios2_type_string, "TestUtilsCWrite");
    adios2_define_attribute_array(ioH, "strarray", adios2_type_string, strarray,
                                  sizeof(strarray) / sizeof(char *));
    adios2_define_attribute_array(ioH, "nwriters", adios2_type_int32_t, &nproc,
                                  1);
    unsigned short shape2D[2] = {(unsigned short)d2_Nx, (unsigned short)d2_Ny};
    adios2_define_attribute_array(ioH, "shape2D", adios2_type_uint16_t, shape2D,
                                  2);
    adios2_define_attribute_array(ioH, "aI8", adios2_type_int8_t, data_I8, 1);
    adios2_define_attribute_array(ioH, "aI16", adios2_type_int16_t, data_I16,
                                  1);
    adios2_define_attribute_array(ioH, "aI32", adios2_type_int32_t, data_I32,
                                  1);
    adios2_define_attribute_array(ioH, "aU8", adios2_type_uint8_t, data_U8, 1);
    adios2_define_attribute_array(ioH, "aU16", adios2_type_uint16_t, data_U16,
                                  1);
    adios2_define_attribute_array(ioH, "aU32", adios2_type_uint32_t, data_U32,
                                  1);
    adios2_define_attribute_array(ioH, "aR32", adios2_type_float, data_R32, 1);
    adios2_define_attribute_array(ioH, "aR64", adios2_type_double, data_R64, 1);

    // inquire variables
    adios2_variable *varNproc = adios2_inquire_variable(ioH, "nproc");
    adios2_variable *varI8 = adios2_inquire_variable(ioH, "varI8");
    adios2_variable *varI16 = adios2_inquire_variable(ioH, "varI16");
    adios2_variable *varI32 = adios2_inquire_variable(ioH, "varI32");
    adios2_variable *varI64 = adios2_inquire_variable(ioH, "varI64");
    adios2_variable *varU8 = adios2_inquire_variable(ioH, "varU8");
    adios2_variable *varU16 = adios2_inquire_variable(ioH, "varU16");
    adios2_variable *varU32 = adios2_inquire_variable(ioH, "varU32");
    adios2_variable *varU64 = adios2_inquire_variable(ioH, "varU64");
    adios2_variable *varR32 = adios2_inquire_variable(ioH, "varR32");
    adios2_variable *varR64 = adios2_inquire_variable(ioH, "varR64");
    adios2_variable *R64_2d = adios2_inquire_variable(ioH, "R64_2d");

    adios2_engine *engineH =
        adios2_open(ioH, "TestUtilsCWriter.bp", adios2_mode_write);

    adios2_put(engineH, varNproc, &nproc, adios2_mode_deferred);
    adios2_put(engineH, varI8, data_I8, adios2_mode_deferred);
    adios2_put(engineH, varI16, data_I16, adios2_mode_deferred);
    adios2_put(engineH, varI32, data_I32, adios2_mode_deferred);
    adios2_put(engineH, varI64, data_I64, adios2_mode_deferred);

    adios2_put(engineH, varU8, data_U8, adios2_mode_deferred);
    adios2_put(engineH, varU16, data_U16, adios2_mode_deferred);
    adios2_put(engineH, varU32, data_U32, adios2_mode_deferred);
    adios2_put(engineH, varU64, data_U64, adios2_mode_deferred);

    adios2_put(engineH, varR32, data_R32, adios2_mode_deferred);
    adios2_put(engineH, varR64, data_R64, adios2_mode_deferred);

    adios2_put(engineH, R64_2d, d2_R64, adios2_mode_deferred);
    adios2_close(engineH);

    // deallocate adiosH
    adios2_finalize(adiosH);

#ifdef ADIOS2_HAVE_MPI
    MPI_Finalize();
#endif

    return 0;
}
