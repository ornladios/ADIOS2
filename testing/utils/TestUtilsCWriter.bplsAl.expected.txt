  int16_t   aI16      attr   = 512
  int32_t   aI32      attr   = 131072
  int8_t    aI8       attr   = 0
  float     aR32      attr   = 0
  double    aR64      attr   = 0
  uint16_t  aU16      attr   = 32768
  uint32_t  aU32      attr   = 2147483648
  uint8_t   aU8       attr   = 128
  string    name      attr   = "TestUtilsCWrite"
  int32_t   nwriters  attr   = 1
  uint16_t  shape2D   attr   = {4, 5}
  string    strarray  attr   = {"first", "second", "third", "fourth"}
